File: server/src/routes/tax-profile.js
// ── /api/tax-profile ────────────────────────────────────────────────────────
// GET    — мой профиль (или дефолт если ещё не сохранён)
// PUT    — сохранить/обновить профиль

const router = require('express').Router();
const db = require('../db');

function requireUser(req, res) {
  const id = req.user?.id;
  if (!id) { res.status(401).json({ error: 'Требуется авторизация' }); return null; }
  return id;
}

const DEFAULT_PROFILE = {
  entity_type: 'ip',
  regime: '910',
  size_category: 'small',
  has_employees: false,
  is_vat_payer: false,
  employees_count: 0,
  annual_revenue: null,
};

router.get('/', async (req, res) => {
  const userId = requireUser(req, res);
  if (!userId) return;
  try {
    const r = await db.query(
      `SELECT entity_type, regime, size_category, has_employees, is_vat_payer,
              employees_count, annual_revenue, updated_at
         FROM company_tax_profile WHERE user_id = $1`,
      [userId]
    );
    if (r.rows.length === 0) return res.json({ ...DEFAULT_PROFILE, exists: false });
    res.json({ ...r.rows[0], exists: true });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

const ALLOWED_ENTITY  = ['ip', 'too', 'individual'];
const ALLOWED_REGIME  = ['esp', 'self_employed', '910', 'oyr', 'retail', null];
const ALLOWED_SIZE    = ['small', 'medium', 'large', null];

router.put('/', async (req, res) => {
  const userId = requireUser(req, res);
  if (!userId) return;
  const {
    entity_type, regime, size_category, has_employees, is_vat_payer,
    employees_count, annual_revenue,
  } = req.body || {};

  if (entity_type && !ALLOWED_ENTITY.includes(entity_type)) {
    return res.status(400).json({ error: `entity_type должен быть из ${ALLOWED_ENTITY.join('/')}` });
  }
  if (regime !== undefined && !ALLOWED_REGIME.includes(regime)) {
    return res.status(400).json({ error: `regime должен быть из ${ALLOWED_REGIME.join('/')} или null` });
  }
  if (size_category !== undefined && !ALLOWED_SIZE.includes(size_category)) {
    return res.status(400).json({ error: `size_category должен быть из ${ALLOWED_SIZE.join('/')} или null` });
  }

  try {
    const r = await db.query(
      `INSERT INTO company_tax_profile
         (user_id, entity_type, regime, size_category, has_employees, is_vat_payer,
          employees_count, annual_revenue)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
       ON CONFLICT (user_id) DO UPDATE
         SET entity_type     = COALESCE(EXCLUDED.entity_type, company_tax_profile.entity_type),
             regime          = EXCLUDED.regime,
             size_category   = EXCLUDED.size_category,
             has_employees   = COALESCE(EXCLUDED.has_employees, company_tax_profile.has_employees),
             is_vat_payer    = COALESCE(EXCLUDED.is_vat_payer, company_tax_profile.is_vat_payer),
             employees_count = COALESCE(EXCLUDED.employees_count, company_tax_profile.employees_count),
             annual_revenue  = COALESCE(EXCLUDED.annual_revenue, company_tax_profile.annual_revenue),
             updated_at      = NOW()
       RETURNING *`,
      [
        userId,
        entity_type || 'ip',
        regime ?? null,
        size_category ?? null,
        !!has_employees,
        !!is_vat_payer,
        employees_count != null ? Number(employees_count) : 0,
        annual_revenue != null ? Number(annual_revenue) : null,
      ]
    );
    res.json({ ...r.rows[0], exists: true });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

module.exports = router;

File: lib/core/models/tax_profile.dart
/// Налоговый профиль компании пользователя.
/// Хранится на бэкенде в `company_tax_profile`, синхронизируется при логине.
/// Используется для рекомендации КБК и подсказок Консультанта.
enum EntityType {
  ip('ip', 'ИП'),
  too('too', 'ТОО'),
  individual('individual', 'Физлицо');

  final String code;
  final String label;
  const EntityType(this.code, this.label);

  static EntityType fromCode(String? c) {
    if (c == 'too') return EntityType.too;
    if (c == 'individual') return EntityType.individual;
    return EntityType.ip;
  }
}

enum TaxRegimeKind {
  esp('esp', 'ЕСП'),
  selfEmployed('self_employed', 'Самозанятый'),
  simplified910('910', 'Упрощёнка (910)'),
  general('oyr', 'ОУР (общеустановленный)'),
  retail('retail', 'Розничный налог');

  final String code;
  final String label;
  const TaxRegimeKind(this.code, this.label);

  static TaxRegimeKind? fromCode(String? c) {
    if (c == null) return null;
    for (final r in TaxRegimeKind.values) {
      if (r.code == c) return r;
    }
    return null;
  }
}

enum SizeCategory {
  small('small', 'Малый бизнес',
      'до 100 сотрудников и до 300 000 МРП оборота в год'),
  medium('medium', 'Средний бизнес',
      'до 250 сотрудников и до 3 000 000 МРП оборота в год'),
  large('large', 'Крупный бизнес',
      'свыше 250 сотрудников или свыше 3 000 000 МРП оборота');

  final String code;
  final String label;
  final String description;
  const SizeCategory(this.code, this.label, this.description);

  static SizeCategory? fromCode(String? c) {
    if (c == null) return null;
    for (final s in SizeCategory.values) {
      if (s.code == c) return s;
    }
    return null;
  }
}

class TaxProfile {
  final EntityType entityType;
  final TaxRegimeKind? regime;
  final SizeCategory? sizeCategory;
  final bool hasEmployees;
  final bool isVatPayer;
  final int employeesCount;
  final num? annualRevenue;
  final bool exists;  // сохранён ли в БД

  const TaxProfile({
    required this.entityType,
    this.regime,
    this.sizeCategory,
    this.hasEmployees = false,
    this.isVatPayer = false,
    this.employeesCount = 0,
    this.annualRevenue,
    this.exists = false,
  });

  factory TaxProfile.fromJson(Map<String, dynamic> j) => TaxProfile(
        entityType: EntityType.fromCode(j['entity_type'] as String?),
        regime: TaxRegimeKind.fromCode(j['regime'] as String?),
        sizeCategory: SizeCategory.fromCode(j['size_category'] as String?),
        hasEmployees: (j['has_employees'] as bool?) ?? false,
        isVatPayer:   (j['is_vat_payer']  as bool?) ?? false,
        employeesCount: (j['employees_count'] is num)
            ? (j['employees_count'] as num).toInt()
            : 0,
        annualRevenue: j['annual_revenue'] is num ? j['annual_revenue'] as num : null,
        exists: (j['exists'] as bool?) ?? true,
      );

  Map<String, dynamic> toJson() => {
        'entity_type':     entityType.code,
        'regime':          regime?.code,
        'size_category':   sizeCategory?.code,
        'has_employees':   hasEmployees,
        'is_vat_payer':    isVatPayer,
        'employees_count': employeesCount,
        'annual_revenue':  annualRevenue,
      };

  TaxProfile copyWith({
    EntityType? entityType,
    TaxRegimeKind? regime,
    Object? sizeCategory = _Sentinel.notSet,
    bool? hasEmployees,
    bool? isVatPayer,
    int? employeesCount,
    Object? annualRevenue = _Sentinel.notSet,
  }) {
    return TaxProfile(
      entityType: entityType ?? this.entityType,
      regime:     regime ?? this.regime,
      sizeCategory: identical(sizeCategory, _Sentinel.notSet)
          ? this.sizeCategory
          : sizeCategory as SizeCategory?,
      hasEmployees:  hasEmployees ?? this.hasEmployees,
      isVatPayer:    isVatPayer ?? this.isVatPayer,
      employeesCount: employeesCount ?? this.employeesCount,
      annualRevenue: identical(annualRevenue, _Sentinel.notSet)
          ? this.annualRevenue
          : annualRevenue as num?,
      exists: exists,
    );
  }

  /// Что писать в подсказке: "ИП на 910", "ТОО (средний бизнес) на ОУР".
  String get humanLabel {
    final parts = <String>[entityType.label];
    if (entityType == EntityType.too && sizeCategory != null) {
      parts.add('(${sizeCategory!.label.toLowerCase()})');
    }
    if (regime != null) parts.add('на ${regime!.label}');
    return parts.join(' ');
  }
}

enum _Sentinel { notSet }
